### PR TITLE
Fix gateway injection when istio.io/rev=<tag>

### LINKED ISSUE #33237: Gateway injection spins up tons of pods
**Bug description**

I set up my istio 1-8-3 using some cmds like below:

```
kubectl apply -f istiod-service.yaml
istioctl install -y -n istio-system -f control-plane.yaml --revision 1-8-3
istioctl install -y -n istio-ingress -f ingress-gateways.yaml --revision 1-8-3
```

```
cat ingress-gateways.yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio-ingress-gw-install
spec:
  profile: empty
  values:
    gateways:
      istio-ingressgateway:
        autoscaleEnabled: false
  components:
    ingressGateways:
    - name: istio-ingressgateway
      namespace: istio-ingress
      enabled: true
      k8s:
        overlays:
        - apiVersion: apps/v1
          kind: Deployment
          name: istio-ingressgateway
          patches:
          - path: spec.template.spec.containers[name:istio-proxy].lifecycle
            value:
               preStop:
                 exec:
                   command: ["sh", "-c", "sleep 5"]

cat istiod-service.yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    app: istiod
    istio: pilot
    release: istio
  name: istiod
  namespace: istio-system
spec:
  type: ClusterIP
  ports:
  - name: grpc-xds
    port: 15010
  - name: https-dns
    port: 15012
  - name: https-webhook
    port: 443
    targetPort: 15017
  - name: http-monitoring
    port: 15014
  selector:
    app: istiod

cat control-plane.yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: control-plane
spec:
  profile: minimal
  meshConfig:
    defaultConfig:
      proxyMetadata:
        ISTIO_META_DNS_CAPTURE: "true"
    enablePrometheusMerge: true
```
Then I install istiod 1-10-0, then used gw injection to deploy canary of the gw.

```
istioctl install -y -n istio-system --revision 1-10-0 -f control-plane.yaml
istioctl x revision tag set canary --revision 1-10-0
# copy cm of istio-1-8-3 and istio-1-10-0 to istio-ingress namespace
kubectl apply -f ../ingress-gateway-injection-canary-image.yaml -n istio-ingress
# this ^^ doesn't create many pods, only 1 pod, failed to reach running.

kubectl label namespace istio-ingress istio.io/rev=canary --overwrite
kubectl apply -f ../ingress-gateway-injection-canary.yaml -n istio-ingress
# this ^^ creates many pods many are able to reach running, many are pending before I kill them all.
```

```
cat ingress-gateway-injection-canary-image.yaml                          (8d0h42m) 16:24:45
apiVersion: apps/v1
kind: Deployment
metadata:
  name: istio-ingressgateway-canary
  namespace: istio-ingress
spec:
  selector:
    matchLabels:
      istio: ingressgateway
      istio.io/rev: canary
  template:
    metadata:
      annotations:
        # Enable gateway injection
        inject.istio.io/templates: gateway
      labels:
        # Set a unique label for the gateway. This is required to ensure Gateways
        # can select this workload
        istio: ingressgateway
        # Enable gateway injection. If connecting to a revisioned control plane, replace with
        istio.io/rev: canary
        # sidecar.istio.io/inject: "true"
    spec:
      containers:
      - name: istio-proxy
        image: docker.io/istio/proxyv2:1.10.0 # The image will automatically update each time the pod starts.
        lifecycle:
          preStop:
            exec:
              command: ["sleep", "5"]

cat ingress-gateway-injection-canary.yaml                                (8d0h42m) 16:25:02
apiVersion: apps/v1
kind: Deployment
metadata:
  name: istio-ingressgateway-canary
  namespace: istio-ingress
spec:
  selector:
    matchLabels:
      istio: ingressgateway
      istio.io/rev: canary
  template:
    metadata:
      annotations:
        # Enable gateway injection
        inject.istio.io/templates: gateway
      labels:
        # Set a unique label for the gateway. This is required to ensure Gateways
        # can select this workload
        istio: ingressgateway
        # Enable gateway injection. If connecting to a revisioned control plane, replace with
        istio.io/rev: canary
        # sidecar.istio.io/inject: "true"
    spec:
      containers:
      - name: istio-proxy
        image: auto # The image will automatically update each time the pod starts.
        lifecycle:
          preStop:
            exec:
              command: ["sleep", "5"]
```

[ ] Docs
[ ] Installation
[x ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

### PR DESCRIPTION
**Pull request by jewertow**
**Please provide a description of this PR:**

**Background context**

This PR restores injecting label `istio.io/rev` to sidecars and fixes #33237 which is still reproducible in Istio 1.17.

Restoring label `istio.io/rev` is not required to fix #33237, but it seems that it was removed unnecessarily in #33447.
I want to restore this label, because it will simplify collecting metrics with revisions. Currently it's not set in sidecars and it requires different relabel configs for sidecars and gateways.

**How to test and reproduce**

1. Prepare files:
```
cat <<EOF >> istiod-service.yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    app: istiod
    istio: pilot
    release: istio
  name: istiod
spec:
  type: ClusterIP
  ports:
  - name: grpc-xds
    port: 15010
  - name: https-dns
    port: 15012
  - name: https-webhook
    port: 443
    targetPort: 15017
  - name: http-monitoring
    port: 15014
  selector:
    app: istiod
EOF
cat <<EOF >> control-plane.yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: control-plane
spec:
  #hub: quay.io/jewertow
  #tag: istio-io-rev
  profile: minimal
  meshConfig:
    defaultConfig:
      proxyMetadata:
        ISTIO_META_DNS_CAPTURE: "true"
    enablePrometheusMerge: true
EOF
cat <<EOF >> ingress-gateway.yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio-ingress-gw-install
spec:
  profile: empty
  #hub: quay.io/jewertow
  #tag: istio-io-rev
  components:
    ingressGateways:
    - name: istio-ingressgateway
      namespace: istio-ingress
      enabled: true
EOF
cat <<EOF >> ingress-gateway-injection-canary.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: istio-ingressgateway-canary
spec:
  selector:
    matchLabels:
      istio: ingressgateway
      istio.io/rev: canary
  template:
    metadata:
      annotations:
        # Enable gateway injection
        inject.istio.io/templates: gateway
      labels:
        # Set a unique label for the gateway. This is required to ensure Gateways
        # can select this workload
        istio: ingressgateway
        # Enable gateway injection. If connecting to a revisioned control plane, replace with
        istio.io/rev: canary
        # sidecar.istio.io/inject: "true"
    spec:
      containers:
      - name: istio-proxy
        image: auto
EOF
```
2. Install:
```
kind create cluster --name test
istioctl install -y -n istio-system -f control-plane.yaml --revision 1-8-3
kubectl apply -n istio-system -f istiod-service.yaml
kubectl create namespace istio-ingress
istioctl install -y -n istio-ingress -f ingress-gateway.yaml --revision 1-8-3
istioctl install -y -n istio-system -f control-plane.yaml --revision 1-10-0
istioctl x revision tag set canary --revision 1-10-0
kubectl apply -n istio-ingress -f ingress-gateway-injection-canary.yaml
```
3. In 1.17 you should see tens or hundreds of pods. Delete them:
```
kubectl delete -n istio-ingress -f ingress-gateway-injection-canary.yaml
kubectl delete pods -l istio=ingressgateway -n istio-ingress
```
4. Now uncomment `hub` and `tag` in control-plane.yaml and ingress-gateway.yaml and use istioctl built from my branch.

### CODE REVIEW

**Comment by jewertow** [comment]
cc: @linsun @Monkeyanator

File: manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: Hmm so the difference between this and the original implementation was before we injected the **actual revision** and now we inject the **desired revision** (which could be a tag).

I like it in some ways (it at least solves the original issue), but don't like it in other ways.

For example, if I have revisions v1 and v2 with tag "canary" and "stable", I may think I can do `kubectl get pods -A -l istio.io/rev=v1` (or look at metrics, since I think you said that was your use case) to find all v1 pods. But this is actually not the case! if a user had `istio.io/rev=canary` which resolves to v1, the pod will also have `istio.io/rev=canary` and we will miss it.

This is not necessarily a bug if we consider the label to mean "desired revision" but I am a bit worried people will assume it is "actual revision"

WDYT @linsun @Monkeyanator

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: > before we injected the actual revision and now we inject the desired revision (which could be a tag).

Yes, you're right, but it doesn't matter from the injection point of view.

I know that users cannot get pods by revision when they use revision tags, but this is the drawback of revision tags, which actually are not required. For me, this is a little downside comparing to consequences of creating pods until resources are exhausted.

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: > Hmm so the difference between this and the original implementation was before we injected the actual revision and now we inject the desired revision (which could be a tag).

Just for clarification: this is necessary to fix #33237, because otherwise tag and revision will never match and it will cause infinite creating pods.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: Yep I get that. I think the tradeoff is no label vs possibly confusing label, not regressing on https://github.com/istio/istio/issues/33237 (which would be bad). I am on the fence, will look for others opinions

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: No, the trade-off is gateway injection works with tag revisions vs confusing label.

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: The change in sidecar injector is not relevant for the bug fix. I should made 2 different PRs.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: I would think the two should be consistent though, right? So if we align with sidecar injector we should remove the label entirely?

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: And they are consistent right now. No, my goal is to have this label on all workloads.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: Sorry I don't get how its consistent with this PR now?

Gateway: inject istio.io/rev label
Sidecars: do not inject istio.io/rev label?

The initial version was consistent since you *also* added it for sidecars, but its not anymore...?

I think the two need to be consistent. I am probably fine with it consistently being "inject istio.io/rev label, with logic here" but i want more feedback from other folks before unilaterally making a change that has wide implications.

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: No, currently gateways have `istio.io/rev` and sidecars don't have. Sidecars had this label in the past, but it was removed in #33447 as a fix for #33237, but in fact it didn't fix the issue with gateway injection. So this change aims to make gateways and sidecars consistent and inject `istio.io/rev` in both cases and at the same time, solve the problem of gateway injection when revision tag is specified.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: I thought you said you were reverting the change to manifests/charts/istio-control/istio-discovery/files/injection-template.yaml. Sorry for the miscommunication.

In that case:

> I am probably fine with it consistently being "inject istio.io/rev label, with logic here" but i want more feedback from other folks before unilaterally making a change that has wide implications.

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: Because I reverted that change to make this consistently. Anyway, if you're fine with this, let's wait for feedback from others.

**Comment by jewertow** [comment]
Something is wrong with the release-notes job, so I will skip it.

File: manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml
Comment (linsun) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: Q: if canary is changed to v3, do we go back to change the pod to be labelled as v3? Or would that only happen if the pod gets restarted?

I think I am leaning towards to be very explicit about the revision for gateways, which would be v2 if that is the revision. Tag is introduced as a convenience for apps, for gateways, I can see users want to be more explicit as it could impact many apps?

**Comment by linsun** [comment]
cc @therealmitchconnors who may also have some input on this

File: manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml
Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: After thinking about this more I am a bit worried about the inconsistency. Let's consider we have a revision R and tag T

If namespace has R, pod gets R
If pod has R, pod gets R
If namespace has T, pod gets R
If pod has T, pod gets T

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: > if canary is changed to v3, do we go back to change the pod to be labelled as v3? Or would that only happen if the pod gets restarted?

If you remove tag `canary` and then create `v3`, a gateway will work until you restart its pod. But a sidecar would behave identically.

> I think I am leaning towards to be very explicit about the revision for gateways

Do you suggest to ignore injection when a gateway is labeled with a tag instead of a revision? What about sidecars then? Sidecars currently work fine when are labeled with `istio.io/rev=<tag>`.

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: The list of inconsistent labeling is valid only for gateways, so in fact currently we have the following rules:
1. Gateways:
- If namespace has R, pod gets R
- If pod has R, pod gets R
- If namespace has T, pod gets R
- If pod has T, pod get R **(and therefore pods fail!)**
2. Sidecars:
- If namespace has R, pod does not get label
- If pod has R, pod gets R
- If namespace has T, pod does not get label
- If pod has T, pod gets T

This PR will change labeling rule as follows:
1. Gateways:
- If namespace has R, pod gets R
- If pod has R, pod gets R
- If namespace has T, pod gets R
- If pod has T, pod get T
2. Sidecars:
- If namespace has R, pod gets R
- If pod has R, pod gets R
- If namespace has T, pod gets R
- If pod has T, pod gets T

So as you can see, this PR makes labeling more consistent that this is for now.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: Right, my suggestion is to make gateways like sidecars are today and not mess with any label. Then in all 4 categories the result is "we don't mess with the label".

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: Making gateways like sidecars are today means that gateway injection supports "If pod has T, pod gets T" as sidecars support it. So that's great, because this is my goal.

If it comes to making gateways like sidecars in case of "If namespace has R/T, pod does not get label", we can't do this, because it would be backward incompatible. Btw. propagating namespace labels is not related to this PR.

If something is really wrong with this change, please point what it is, because I completely don't understand your position to this PR. I solved a bug which looks major or critical, because it directly affects performance of kube-apiserver and etcd and you block this, because you worry what labeling rules users may expect while we even don't have data to draw conclusions. In fact, we just suppose while it may not be a problem at all, and at the same time, it makes collecting metrics per revision more complicated. I don't understand why do you ignore this goal.

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: > The only way I see to achieve those requirements and meet your goal is to remove the injected istio.io/rev label on gateway pods (no change needed to sidecar as it was already removed).

Removing injection of `istio.io/rev` to pods prevents collecting metrics per revision, so what do you think about changing case "If namespace has T, pod gets R" to "If namespace has T, pod gets T"?

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: "If namespace has T, pod gets T"?

How will this be implemented? We don't have access to this info currently in the injection

I still don't think it meets your goal of "metrics per revision" since it's *not the revision*, it's the revision OR the tag which makes it not useful for most purposes

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: I took a look how `revision tag set` works and I guess that it should be feasible to create webhook so that it knows what should be injected to istio.io/rev. If setting revision based on tag is possible I guess that setting tag should be as well.

> I still don't think it meets your goal of "metrics per revision"

This is true, but I don't want to remove the rule "If namespace has R, pod gets R" and therefore I care about the case with tag. It's not so simple to keep "If namespace has R, pod gets R" and removing "If namespace has T, pod gets R". This would require an additional refactor in injection webhooks.

I think that the best now is to narrow this PR to just fix gateway injection issue and I will address the problem of propagating namespace label to pods in a subsequent PR in which I will try to implement "If namespace has T, pod gets T".

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: Just to clarify, I am not sure "If namespace has T, pod gets T" is appropriate either. It's highly confusing to have a label injected, for purpose of metrics or identifying revision usage etc when they it's NOT actually recording the revision. 

IMO there are 2 valid paths:
* remove the istio.io/rev label injection in gateway (align with sidecar)
* Make a new label (strawman: istio.io/actual-revision). Always set this to .Revision. Reject pods that have this label already set to avoid the Deployment selector issue (except.. this may break idempotency and not be viable...). This is like the old logic but attempting to avoid the Deployment issue

The latter would be a new API and I'm not sure it's a good idea at this point (nor am I sure it actually avoids the selector issue safely). The former I will certainly approve, although I don't think it meets your use cases(?).

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: As mentioned above I am not okay with the change here. See the options in https://github.com/istio/istio/pull/43433/files#r1112378489.

To be clear - I acknowledge you are fixing a bug here. However, "fixes a bug" is not the only criteria to accept a PR. This introduces a regression in a variety of ways (understability, semantics, consistency).

How about we just change this to remove the istio.io/rev label injection?

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: Please look at this once again. Currently, label injection works as follows:

Gateways:
1. If namespace has R, pod gets R
2. If pod has R, pod gets R
3. If namespace has T, pod gets R
4. **If pod has T, pod get R (injection fails!)**

Sidecars:
1. If namespace has R, pod does not get label
2. If pod has R, pod gets R
3. If namespace has T, pod does not get label
4. If pod has T, pod gets T

After reverting changes made in sidecar injection template, labeling will work as follows:

Gateways:
1. If namespace has R, pod gets R
2. If pod has R, pod gets R
3. If namespace has T, pod gets R
4. **If pod has T, pod get T**

Sidecars: **(rules didn't change**)
1. If namespace has R, pod does not get label
2. If pod has R, pod gets R
3. If namespace has T, pod does not get label
4. If pod has T, pod gets T

As you can see, I changed only the failing scenario, which is consistent with current sidecar injection and I didn't introduce any regression, understability, etc.

> How about we just change this to remove the istio.io/rev label injection?

As I already said, I really want keep case "If namespace has R, pod gets R", because we will really need this to support multi-revision observability. I know that you suggested alternatives, but those also require analysis of pros and cons, and I am not sure if we really need to remove `istio.io/rev` in favor of another label. I don't want to change this PR back and forth. I will address that problem in a follow-up PR and I will take into account your expectations.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: When you phrase it like that (which, admittedly, was how I phrased it :-) ), I agree it makes Sidecar and Gateway the same with this change. however, the missing case is

"If pod has no label"

Gateways with this change: "If pod has no istio.io/rev label, gets R"
Sidecars: "If pod has no istio.io/rev label, gets no label"

That is the inconsistency... the `istio.io/rev` has different meanings at different times.

The sidecar case as today could be simplified to "The pod label is not mutated". The gateway case cannot.

---

Its not clear to me why you need multi-revision observability *only* for gateways..? Is that actually the case? If not, why do we want them to behave differently.

Comment (jewertow) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: But I don't try to address case "If pod has no label". The point is that "If pod has label T, gateway injection fails!", and current scope of this PR addresses this case. Nothing else, because other cases work. I did a mistake trying to address to different things in a single PR, but I already limited scope and updated title.

If it comes to `Gateways with this change: "If pod has no istio.io/rev label, gets R"` - this is implied by "If namespace has R/T, pod get R/T". But please note that I didn't add this - I just didn't remove this, because I want to fix the bug keeping backward compatibility.

> Its not clear to me why you need multi-revision observability only for gateways..?

We need multi-revision observability for gateways as well as for sidecars. And this is why I initially reverted changes made in the #33447, which removed the rule "If namespace has R, pod gets R", but then reverted this from my PR, because of your objections. Removing istio.io/rev entirely, removes the rule "If namespace has R, pod gets R" which we really need for multi-revision observability. Otherwise, this will be much less usable.

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml:7]: > But I don't try to address case "If pod has no label"

And I am saying we cannot ignore these cases...

We are going in circles, lets discuss in the WG meeting.

**Comment by rcernich** [comment]
Hey @howardjohn, could you please summarize what you're looking for.  I've lost my hold on the thread.  TIA

**Comment by howardjohn** [comment]
I want the istio.io/rev label to be removed instead of modified

**Comment by jewertow** [comment]
I already explained that we need this label for multi-revision observability. If a namespace will not propagate label to pods, this feature will not make sense, because will require to label every single workload.

**Comment by howardjohn** [comment]
My concern is there is no consistency with sidecar (which used to have this label but **removed** it, as I am suggesting here, in https://github.com/istio/istio/pull/33447), which is not acceptable. I am additionally concerned that the meaning of `istio.io/rev` becomes unclear - in particular, I don't think your goal of monitoring revisions is actually met by this implementation, as we are NOT putting the "revision" in the label in some case.

I am aware of your requirement to monitor revisions, but its I believe this PR is inappropriate both in general and as a mechanism to achieve that goal. To come up with alternatives it would help to understand your requirements a bit more. For example, does an annotation work? Is this prometheus scraping or some other system? etc

**Comment by jewertow** [comment]
I didn't introduce this inconsistency. This is an existing rule. And since we cannot find a common point of view on multi-revision observability, I don't want to address this use case in this PR. I just want to fix an existing bug, backport it to previous affected releases and that's it. I will revise current labeling rules and then we can talk on WG meeting how we can solve inconsistent labeling (which was not introduced in this PR) and how to support multi-revision observability. I will take into account your expectations.

**Comment by howardjohn** [comment]
> I didn't introduce this inconsistency. This is an existing rule.

Just because it's currently inconsistent doesn't mean we should make it worse. I would be happy to open a PR to resolve the inconsistency if you would not like to do so in this PR

**Comment by jewertow** [comment]
What did I make worse in this PR? I fixed only the case that does not work and which is not supposed to be not supported.

**Comment by jewertow** [comment]
I won't be able to backport this fix to previous release if I will extend the scope... Do I really have to please you to approve a bug fix? Isn't it enough that we invested time in fixing a bug that wasn't solved since more than a year? And now I cannot get an approval, because you just want to address something else **additionally**? This is a way how we work as a community?

**Comment by howardjohn** [comment]
Let me address those one-by-one

> I won't be able to backport this fix to previous release if I will extend the scope...

Lets first focus on making a proper fix, and if that expands the scope too much we can discuss a smaller scoped PR to backport.

To be 100% clear the fix I am suggesting is removing L7 of manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml, so it is no larger in scope than the current PR (IMO, anyways)

> Do I really have to please you to approve a bug fix? 

No, you do not. However, you do need to get approval from an environments WG member, as they are owners of this folder; I happen to be one but am not the only one. When a maintainer and contributor cannot come to a consensus, the next step is to bring to the WG where we can discuss, and if necessary bring to a vote. If there cannot be consensus in the WG, it can be further escalated to the TOC.

These policies are documented in https://github.com/istio/community. As our community governance is also open, you are welcome to propose changes to those policies as well (although that would, of course, involve more discussion than a simple PR).


> Isn't it enough that we invested time in fixing a bug that wasn't solved since more than a year?

No; fixing a bug is an excellent contribution and one I am happy to see, it is not sufficient to ensure a PR is merged. There are a variety of factors involved when approving PRs, involving API design, consistency, performance, complexity, safety, etc.

> And now I cannot get an approval, because you just want to address something else **additionally**?

Specifically, my concern is that you are moving from a <not desired state> to a <not desired state, but in different ways>. The PR honestly might be an improvement to the current state (its a bit of a tradeoff, but it probably is as the bug is pretty severe when triggered), but I am more comparing to the actual optimal (IMO) state of simply removing L7.

In general, I do think that expanding PR scopes to be "perfect" instead of approving PRs that are small but bring improvement, and often advocate for making forward progress. However, this is not the case on this PR, where the ideal fix is easier than the proposed fix.

> This is a way how we work as a community?

I believe this was addressed above, let me know if I can help clarify.

Hope that helps address some of your concerns.

**Comment by jewertow** [comment]
Ok, so before we will start to talk about removing/keeping istio.io/rev on WG meeting, please explain what's technically wrong with this label - assuming that the case "If namespace has T, pod gets R" would not be supported.

**Comment by jewertow** [comment]
You also still didn't explain what you mean by making it worse, regression, understability, etc. If you are not going to explain what you mean by that, and give examples how the solution breaks something, then it's not a constructive feedback.

Don't be surprised that I am disappointed in your approach, but you had completely different view on changing existing labels in #43388, and in this case you want to remove the label and don't care about users, like we, which want to use `istio.io/rev`.

Anyway, this discussion does not make sense anymore, because I see that you already decided that you will remove this label independently on our use case, so let's consider another trade-off.

You suggested to remove `istio.io/rev` and inject `istio.io/actual-revision`, which would be always set to `.Revision`, but a user may still make a mistake and apply a Deployment with `matchLabel` including `istio.io/actual-revision` what will result in indefinite creating pods, like it is for now. Other labels in injection templates, like `service.istio.io/canonical-name` or `service.istio.io/canonical-revision` are not overwritten if are already set, so I don't think that `istio.io/actual-revision` must be overwritten. I would agree to remove `istio.io/rev` and replace with non-overwritten `istio.io/actual-revision` in all components: sidecars, default gateways and injected gateways.

**Comment by hzxuzhonghu** [comment]
@jewertow This topic is added in the agenda [[jewertow](https://github.com/jewertow)](https://docs.google.com/document/d/1wsa06GGiq1LEGwhkiPP0FKIZJqdAiue-VeBonWAzAyk/edit#) by @howardjohn 
But since you are not there, they do not discuss this topic, would you be able to attend next week

**Comment by hzxuzhonghu** [comment]
cc @istio/wg-environments-maintainers

**Comment by jewertow** [comment]
Thanks @hzxuzhonghu for letting me know.

**Review by howardjohn — APPROVED** [review]
LGTM, thanks for the fix. I am interested in the use case of monitoring revisions, feel free to followup to discuss how we can better enable this (for gateways and sidecars)

**Comment by jewertow** [comment]
I removed cherry-pick labels, because releases 1.15, 1.16, 1.17 will require backward compatible fix. I will submit proper pull requests later.

**Comment by jewertow** [comment]
@howardjohn could you remove "do-not-merge/hold" label since you approved this PR?

**Pull request merged**
